app/master/build.py

```diff
@@ -309,6 +288,42 @@ def _on_enter_error_state(self, event):
         self._error_message = getattr(event, 'error_msg', default_error_msg)
         self._logger.warning('Build {} failed: {}', self.build_id(), self._error_message)
 
+    def _on_enter_preparing_state(self, event):
+        """
+        Prepare the build by atomization and subjobs creation.
+        This method is triggered by a state machine transition to the PREPARING state.
+        :param event: The Fysom event object
+        :type event: BuildEvent
+        """
+        self._logger.info('Fetching project for build {}.', self._build_id)
+        self.project_type.fetch_project()
+        self._logger.info('Successfully fetched project for build {}.', self._build_id)
+
+        job_config = self.project_type.job_config()
+        if job_config is None:
+            raise RuntimeError('Build failed while trying to parse clusterrunner.yaml.')
+
+        subjob_calculator = getattr(event, 'subjob_calculator', None)
+        if subjob_calculator is None:
+            raise RuntimeError('Build failed as subjob_calculator is not set for this build {}.', self._build_id)
+        try:
+            subjobs = subjob_calculator.compute_subjobs_for_build(self._build_id, job_config, self.project_type)
+        except AtomizerError as ex:
+            if self.is_canceled:
+                # Add info to the AtomizerError error message. If build is canceled, then AtomizerError is expected.
+                ex.args = ex.args + ('This is expected as Build {} is CANCELED.'.format(self._build_id),)
```

Comment:
Is this exposed anywhere if the build is canceled? It looks like the exception message is only used if the build is not canceled, so maybe this extra message (and the try/except) is not needed.